osquery/distributed/plugins/bro_distributed.cpp

```diff
@@ -0,0 +1,336 @@
+  LOG(INFO) << "Received event '" << eventName << "' on topic '" << topic
+            << "'";
+
+  // osquery::host_execute
+  if (eventName == bm.EVENT_HOST_EXECUTE) {
+    // One-Time Query Execution
+    SubscriptionRequest sr;
+    createSubscriptionRequest("EXECUTE", msg, topic, sr);
+    std::string newQID = qm.addOneTimeQueryEntry(sr);
+    if (newQID.empty()) {
+      return Status(8, "Unable to add Broker Query Entry");
+    }
+    DistributedQueryRequest dqr;
+    dqr.id = newQID;
+    dqr.query = sr.query;
+    oT_queries.push_back(dqr);
+    return Status(0, "OK");
+
+    // osquery::host_join
+  } else if (eventName == bm.EVENT_HOST_JOIN) {
+    std::string newGroup = *broker::get<std::string>(msg[1]);
+    return bm.addGroup(newGroup);
+
+    // osquery::host_leave
+  } else if (eventName == bm.EVENT_HOST_LEAVE) {
+    std::string newGroup = *broker::get<std::string>(msg[1]);
+    return bm.removeGroup(newGroup);
+
+    // osquery::host_subscribe
+  } else if (eventName == bm.EVENT_HOST_SUBSCRIBE) {
+    // New SQL Query Request
+    SubscriptionRequest sr;
+    createSubscriptionRequest("SUBSCRIBE", msg, topic, sr);
+    Status s_sub = qm.addScheduleQueryEntry(sr);
+    if (!s_sub.ok()) {
+      return s_sub;
+    }
+
+    // osquery::host_unsubscribe
+  } else if (eventName == bm.EVENT_HOST_UNSUBSCRIBE) {
+    // SQL Query Cancel
+    SubscriptionRequest sr;
+    createSubscriptionRequest("UNSUBSCRIBE", msg, topic, sr);
+    // TODO: find an UNIQUE identifier (currently the exact sql string)
+    std::string query = sr.query;
+
+    Status s_unsub = qm.removeQueryEntry(query);
+    if (!s_unsub.ok()) {
+      return s_unsub;
+    }
+
+  } else if (eventName == "osquery::host_test") {
+  } else {
+    // Unkown Message
+    return Status(7, "Unknown event name '" + eventName + "'");
+  }
+
+  // Apply to new config/schedule
+  std::map<std::string, std::string> config_schedule;
+  config_schedule["bro"] = qm.getQueryConfigString();
+  VLOG(1) << "Applying new schedule: " << config_schedule["bro"];
+  Config::get().update(config_schedule);
+
+  return Status(0, "OK");
+}
+
+Status BRODistributedPlugin::getQueries(std::string& json) {
+  BrokerManager& bm = BrokerManager::get();
+
+  // Collect file descriptors of the broker message queues
+  fd_set fds;
+  FD_ZERO(&fds);
+  std::vector<std::string> topics = bm.getTopics(); // List of subscribed topics
+  int sMax = 0;
+  // Retrieve info about each message queue
+  for (const auto& topic : topics) {
+    int sock = bm.getMessageQueue(topic)->fd();
+    if (sock > sMax) {
+      sMax = sock;
+    }
+    FD_SET(sock, &fds); // each topic -> message_queue -> fd
+  }
+  // Append the connection status file descriptor to detect connection failures
+  FD_SET(bm.getOutgoingConnectionFD(), &fds);
+  sMax = std::max(sMax, bm.getOutgoingConnectionFD());
+
+  // Wait for incoming message
+  int select_code = select(sMax + 1, &fds, nullptr, nullptr, nullptr);
```

Comment:
Use `poll` instead, we had a nasty issue with `select` as we recommend increasing FD max past 1024, and make use of `inotify` persistent FDs so you may overflow your `fd_set`.